src/nunavut/lang/c/templates/_composite_type.j2

```diff
@@ -90,4 +90,54 @@ inline void {{ composite_type | full_reference_name }}_init({{ composite_type |
  */
 #define {{ composite_type | full_reference_name }}_MAX_SERIALIZED_REPRESENTATION_SIZE_BYTES ({{ composite_type.bit_length_set | max }}U)
 
+// Blah blah little-endian has better copying
+// blah blah
+#ifndef CANARD_DSDL_CONFIG_LITTLE_ENDIAN
+#    define CANARD_DSDL_CONFIG_LITTLE_ENDIAN false
+#endif
+
+/**
+ * Internal serialization of {{ composite_type | full_reference_name }} instance into a byte array.
+ *
+ * @param in_instance   A structure instance to serialize.
+ * @param offset        Offset (in bits) of data type in message.
+ * @param out_buffer    Output message buffer.
+ *
+ * @returns             Size of encoded data (in bits)
+ */
+inline size_t _{{ composite_type | full_reference_name }}_serialize({{ composite_type | full_reference_name }}* in_instance, size_t offset, uint8_t* out_buffer)
+{
+    size_t length = 0;
+
+    {#- Error checking concerns:
+        - if lengths of arrays are out of bounds? idk if it's something we need to care about
+          the base data type only has a certain length anyways, so it's hard to go out of bounds
+    #}
```

Comment:
This here. I'm not concerned. It's a bit of a trade off between performance and being hand-holdy IMO, but there is an error case here that isn't checked at the moment.

Reply:
```
inline int32_t _{{ composite_type | full_reference_name }}_serialize({{ composite_type | full_reference_name }}* in_instance, size_t offset, uint8_t* out_buffer, uint32_t* out_length_bits)
```

return 0 for no errors and error codes we define in a support header that are < 0. The size of the encoded data is then (optionally) returned as an out argument.

size_t isn't appropriate since this is sized based on the maximum addressable location for a system in *bytes*. By using uint32_t we avoid weird limitations on, say, AVR8.